plugins/logitech-tap/fu-logitech-tap-sensor-device.c

```diff
@@ -0,0 +1,371 @@
+	if (!fu_logitech_tap_sensor_device_set_feature(self,
+						      (guchar *)set_data,
+							  HID_SET_DATA_LEN,
+						      error))
+	return FALSE;
+	if (!fu_logitech_tap_sensor_device_get_feature(self,
+						      (guchar *)get_data,
+							  HID_GET_DATA_LEN,
+						      error))
+	return FALSE;
+
+	/* little-endian data. MinorVersion byte 3, MajorVersion byte 4, BuildVersion byte 2 & 1 */
+	fwversion =
+	    (get_data[4] << 24) + (get_data[3] << 16) + (get_data[2] << 8) + get_data[1];
+	fu_device_set_version_from_uint32(device, fwversion);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_tap_sensor_device_set_serial(FuDevice *device, GError **error)
+{
+	FuLogitechTapSensorDevice *self = FU_LOGITECH_TAP_SENSOR_DEVICE(device);
+	g_autoptr(GString) serial_number = g_string_new(NULL);
+	guint8 set_data[HID_SET_DATA_LEN] = {kHidMcuCmdSetSerialNumber, kHidMcuSerialNumberSetReportByte1,
+	kHidMcuSerialNumberSetReportByte2, kHidMcuSerialNumberSetReportByte3, kHidMcuSerialNumberSetReportByte4};
+
+	g_debug("get sensor serial number");
+
+	/* setup HID report for serial number */
+	if (!fu_logitech_tap_sensor_device_set_feature(self,
+						      (guchar *)set_data,
+							  HID_SET_DATA_LEN,
+						      error))
+	return FALSE;
+	fu_device_sleep(FU_DEVICE(self), kLogiDefaultSensorSleepIntervalMs); /* 50 ms */
+ 	/* serial number is a 12-byte-string that is stored in MCU  */
+ 	/* each get request fetchs 1 word (4 bytes), so iterate 3 times */
+    for (int index = 1; index <= 3; index++) {
+	    guint8 get_data[HID_GET_DATA_LEN] = {kHidMcuCmdGetSerialNumber, 0, 0, 0, 0};
+	    if (!fu_logitech_tap_sensor_device_get_feature(self,
+						      (guchar *)get_data,
+							  HID_GET_DATA_LEN,
+						      error))
+	        return FALSE;
+	    g_string_append_printf(serial_number, "%c%c%c%c", (guchar)get_data[1],
+	       (guchar)get_data[2], (guchar)get_data[3], (guchar)get_data[4]);
+	}
+
+    fu_device_set_serial(device, serial_number->str);
+
+	/* success */
+	return TRUE;
+}
+
+
+static gboolean
+fu_logitech_tap_sensor_device_setup(FuDevice *device, GError **error)
+{
+	if (!fu_logitech_tap_sensor_device_set_version(device, error))
+		return FALSE;
+	/* enable TDE mode */
+	if (!fu_logitech_tap_sensor_device_set_tde(device, kHidMcuTdeModeEnable, error))
+		return FALSE;
+	/* get serial number */
+	if (!fu_logitech_tap_sensor_device_set_serial(device, error)) {
+		/* disable TDE mode */
+		fu_logitech_tap_sensor_device_set_tde(device, kHidMcuTdeModeDisable, error);
+		return FALSE;
+	}
+	/* disable TDE mode */
+	if (!fu_logitech_tap_sensor_device_set_tde(device, kHidMcuTdeModeDisable, error))
+		return FALSE;
+
+	/* setup device identifier so plugin can disntiguish device during composite_cleaup */
+	fu_device_add_private_flag(device, FU_LOGITECH_TAP_DEVICE_TYPE_SENSOR);
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_tap_sensor_device_probe(FuDevice *device, GError **error)
+{
+#if G_USB_CHECK_VERSION(0, 3, 3)
```

Comment:
This check doesn't seem right, what are you using in gusb?

Reply:
Removed